File: app/src/main/java/com/saienko/androidthings/barman/ui/settings/AddPositionActivity.java
package com.saienko.androidthings.barman.ui.settings;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.widget.*;
import com.google.android.things.pio.Gpio;
import com.google.android.things.pio.PeripheralManagerService;
import com.saienko.androidthings.barman.R;
import com.saienko.androidthings.barman.Utils;
import com.saienko.androidthings.barman.db.DatabaseUtil;
import com.saienko.androidthings.barman.db.cocktail.Component;
import com.saienko.androidthings.barman.db.position.Position;
import com.saienko.androidthings.barman.ui.base.BaseActivity;
import com.saienko.androidthings.barman.ui.settings.adapter.SelectComponentAdapter;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

import java.io.IOException;
import java.util.List;

public class AddPositionActivity extends BaseActivity {

    private static final String TAG = "AddPositionActivity";

    public static final  String EXTRA_POSITION = "EXTRA_POSITION";
    private static final String EXTRA_EDIT     = "EXTRA_EDIT";
    private Position                 position;
    private List<Component>          componentList;
    private Component                selectedComponent;
    private boolean                  isNew;
    private Spinner                  spinnerComponent;
    private PeripheralManagerService pioService;
    private Gpio                     mLedGpio;
    private SelectComponentAdapter   adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_position);
        handleIntent();
        initUI();
        if (Utils.isThingsDevice(this)) {
            pioService = new PeripheralManagerService();
        }
    }

    public static void add(BaseActivity activity, Position position, int regCode) {
        Intent intent = new Intent(activity, AddPositionActivity.class);
        intent.putExtra(EXTRA_POSITION, position);
        activity.startActivityForResult(intent, regCode);
    }

    public static void edit(BaseActivity activity, Position position, int regCode) {
        Intent intent = new Intent(activity, AddPositionActivity.class);
        intent.putExtra(EXTRA_POSITION, position);
        intent.putExtra(EXTRA_EDIT, true);
        activity.startActivityForResult(intent, regCode);
    }

    private void handleIntent() {
        position = getIntent().getParcelableExtra(EXTRA_POSITION);
        isNew = !getIntent().hasExtra(EXTRA_EDIT);
    }

    @Override
    public void initUI() {
        super.initUI();
        TextView tvMotor = findViewById(R.id.tvMotor);
        spinnerComponent = findViewById(R.id.spinnerComponent);
        Button btnSave = findViewById(R.id.btnSave);
        Button btnPour = findViewById(R.id.btnPour);

        btnPour.setOnTouchListener((view, motionEvent) -> {
            switch (motionEvent.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    btnPour.setBackgroundColor(getColor(R.color.colorAccent));
                    startPour(position);
                    break;
                case MotionEvent.ACTION_UP:
                case MotionEvent.ACTION_CANCEL:
                    btnPour.setBackgroundColor(getColor(R.color.lighter_gray));
                    endPour();
                    break;
                default:
                    break;
            }
            return false;
        });

        tvMotor.setText(position.getMotor().getMotorName());
        spinnerComponent.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                selectedComponent = componentList.get(i);
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {
                // IGNORE
            }
        });

        btnSave.setOnClickListener(view -> {
            position.setComponent(selectedComponent);
            savePosition(position);
        });

        if (isNew) {
            Single.fromCallable(DatabaseUtil::getFreeComponents).subscribeOn(Schedulers.io())
                  .observeOn(AndroidSchedulers.mainThread())
                  .subscribe(components -> {
                      componentList = components;
                      spinnerComponent.setAdapter(getAdapter());
                  });
        } else {
            Single.fromCallable(DatabaseUtil::getComponents).subscribeOn(Schedulers.io())
                  .observeOn(AndroidSchedulers.mainThread())
                  .subscribe(components -> {
                      componentList = components;
                      spinnerComponent.setAdapter(getAdapter());
                      if (!isNew) {
                          spinnerComponent.setSelection(getPosition(componentList, position.getComponent().getId()));
                      }
                  });
        }
    }

    private SpinnerAdapter getAdapter() {
        if (adapter == null) {
            adapter = new SelectComponentAdapter(AddPositionActivity.this, componentList);
        }
        return adapter;
    }

    private void endPour() {
        Log.d(TAG, "endPour: ");
        if (pioService != null && mLedGpio != null) {
            try {
                mLedGpio.close();
            } catch (IOException e) {
                Log.e(TAG, "Error closing LED GPIO", e);
            } finally {
                mLedGpio = null;
            }
            mLedGpio = null;
        }
    }

    private void startPour(Position position) {
        Log.d(TAG, "startPour: ");
        if (pioService != null && mLedGpio == null) {
            try {
                mLedGpio = pioService.openGpio(position.getMotor().getGpio().getGpioName());
                mLedGpio.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW);
            } catch (IOException e) {
                Log.e(TAG, "startPour: ", e);
            }
        }

    }

    private void savePosition(Position position) {
        Intent intent = new Intent();
        intent.putExtra(EXTRA_POSITION, position);
        setResult(RESULT_OK, intent);
        finish();
    }

    private int getPosition(List<Component> components, long componentId) {
        for (int i = 0; i < components.size(); i++) {
            if (components.get(i).getId() == componentId) {
                return i;
            }
        }
        return 0;
    }

}

File: app/src/main/java/com/saienko/androidthings/barman/ui/settings/AddPositionActivity.kt
package com.saienko.androidthings.barman.ui.settings

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.MotionEvent
import android.view.View
import android.widget.AdapterView
import android.widget.Button
import android.widget.Spinner
import android.widget.TextView
import com.google.android.things.pio.Gpio
import com.google.android.things.pio.PeripheralManager
import com.saienko.androidthings.barman.R
import com.saienko.androidthings.barman.Utils
import com.saienko.androidthings.barman.db.entity.Component
import com.saienko.androidthings.barman.db.entity.Position
import com.saienko.androidthings.barman.db.interfaces.IComponentManager
import com.saienko.androidthings.barman.db.interfaces.IPositionManager
import com.saienko.androidthings.barman.db.manager.ComponentManager
import com.saienko.androidthings.barman.db.manager.PositionManager
import com.saienko.androidthings.barman.ui.base.BaseActivity
import com.saienko.androidthings.barman.ui.settings.adapter.SelectComponentAdapter
import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.launch
import java.io.IOException

class AddPositionActivity : BaseActivity() {
    private var currentPosition: Position? = null
    private lateinit var componentList: List<Component>
    private var selectedComponent: Component? = null

    private var isNew: Boolean = false
    private var manager: PeripheralManager? = null
    private var gpio: Gpio? = null

    private lateinit var localAdapter: SelectComponentAdapter

    private lateinit var componentManager: IComponentManager
    private lateinit var positionManager: IPositionManager

    private lateinit var spinnerComponent: Spinner
    private lateinit var tvMotor: TextView
    private lateinit var btnSave: Button
    private lateinit var btnPour: Button

    private var positionId: Long = -1
    private var motorId: Long = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_position)
        componentManager = ComponentManager()
        positionManager = PositionManager()
        initUI()
        handleIntent()
        if (Utils.isThingsDevice(this)) {
            manager = PeripheralManager.getInstance()
        }
    }

    private fun handleIntent() {
        positionId = intent.getLongExtra(EXTRA_POSITION_ID, positionId)
        isNew = !intent.hasExtra(EXTRA_EDIT)
        if (isNew) {
            motorId = intent.getLongExtra(EXTRA_MOTOR_ID, motorId)
        }
        getPosition(positionId)
    }

    private fun getPosition(positionId: Long) {
        launch(UI) {
            initPosition(async(CommonPool) { positionManager.get(positionId) }.await())
        }
    }

    private fun initPosition(position: Position) {
        if (position.motor != null) {
            tvMotor.text = position.motor!!.motorName
        } else {
            tvMotor.setText(R.string.unknown)
            btnSave.isEnabled = false
            btnPour.isEnabled = false
        }
        currentPosition = position

        launch(UI) {
            val list: MutableList<Component> = async(CommonPool) {
                if (isNew) {
                    componentManager.allFree()
                } else {
                    val list = componentManager.allFree()
                    position.component?.let { list.add(it) }
                    return@async list
                }
            }.await()
            showPosition(list, position)
        }
    }

    private fun showPosition(
            components: MutableList<Component>,
            position: Position) {
        componentList = components
        initAdapter()
        spinnerComponent.adapter = localAdapter
        if (!isNew) {
            spinnerComponent.setSelection(
                    getPosition(componentList, position.component!!.id))
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun initUI() {
        super.initUI()
        tvMotor = findViewById(R.id.tvMotor)
        spinnerComponent = findViewById(R.id.spinnerComponent)
        btnSave = findViewById(R.id.btnSave)
        btnPour = findViewById(R.id.btnPour)

        btnPour.setOnTouchListener { _, motionEvent ->
            when (motionEvent.action) {
                MotionEvent.ACTION_DOWN -> {
                    btnPour.setBackgroundColor(getColor(R.color.colorAccent))
                    startPour(currentPosition)
                }
                MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                    btnPour.setBackgroundColor(getColor(R.color.lighter_gray))
                    endPour()
                }
                else -> {
                }
            }
            false
        }

        spinnerComponent.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>, view: View, i: Int, l: Long) {
                selectedComponent = componentList[i]
                if (selectedComponent != null) {
                    btnSave.isEnabled = true
                }
            }

            override fun onNothingSelected(adapterView: AdapterView<*>) {
                // IGNORE
            }
        }

        btnSave.setOnClickListener {
            if (isNew) {
                currentPosition = Position(selectedComponent!!.id, motorId)
                savePosition(currentPosition!!)
            } else {
                currentPosition!!.component = selectedComponent
                updatePosition(currentPosition!!)
            }
        }
    }

    private fun savePosition(position: Position) {
        launch(UI) {
            async(CommonPool) { positionManager.insert(position) }.await()
            finish()
        }
    }

    private fun updatePosition(position: Position) {
        launch(UI) {
            async(CommonPool) { positionManager.update(position) }.await()
            finish()
        }
    }

    private fun initAdapter() {
        localAdapter = SelectComponentAdapter(this@AddPositionActivity, componentList)
    }

    private fun endPour() {
        if (manager != null && gpio != null) {
            try {
                gpio!!.close()
            } catch (e: IOException) {
                Log.e(TAG, "Error closing LED GPIO", e)
            } finally {
                gpio = null
            }
            gpio = null
        }
    }

    private fun startPour(position: Position?) {
        if (manager != null && gpio == null) {
            try {
                gpio = manager!!.openGpio(position!!.motor!!.gpio!!.gpioName)
                gpio!!.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW)
            } catch (e: IOException) {
                Log.e(TAG, "startPour: ", e)
            }
        }
    }

    private fun getPosition(components: List<Component>, componentId: Long): Int {
        return components.indices.firstOrNull { components[it].id == componentId }
               ?: -1
    }

    companion object {

        const val EXTRA_POSITION_ID = "EXTRA_POSITION_ID"
        private const val TAG = "AddPositionActivity"
        private const val EXTRA_EDIT = "EXTRA_EDIT"
        private const val EXTRA_MOTOR_ID = "EXTRA_MOTOR_ID"

        fun add(context: Context, positionId: Long, motorId: Long) {
            val intent = Intent(context, AddPositionActivity::class.java)
            intent.putExtra(EXTRA_POSITION_ID, positionId)
            intent.putExtra(EXTRA_MOTOR_ID, motorId)
            context.startActivity(intent)
        }

        fun edit(context: Context, positionId: Long) {
            val intent = Intent(context, AddPositionActivity::class.java)
            intent.putExtra(EXTRA_POSITION_ID, positionId)
            intent.putExtra(EXTRA_EDIT, true)
            context.startActivity(intent)
        }
    }
}

